lib/datadog/tracing/sampling/span/rule_parser.rb

```diff
@@ -0,0 +1,98 @@
+# frozen_string_literal: true
+
+require 'json'
+require 'datadog/tracing/sampling/span/ext'
+require 'datadog/tracing/sampling/span/matcher'
+require 'datadog/tracing/sampling/span/rule'
+
+module Datadog
+  module Tracing
+    module Sampling
+      module Span
+        # Converts user configuration into {Datadog::Tracing::Sampling::Span::Rule} objects,
+        # handling any parsing errors.
+        module RuleParser
+          class << self
+            # Parses the provided JSON string containing the Single Span
+            # Sampling configuration list.
+            # In case of parsing errors, `nil` is returned.
+            #
+            # @param rules [String] the JSON configuration rules to be parsed
+            # @return [Array<Datadog::Tracing::Sampling::Span::Rule>] a list of parsed rules
+            # @return [nil] if parsing failed
+            def parse_json(rules)
+              begin
+                list = JSON.parse(rules)
+              rescue JSON::ParserError => e
+                Datadog.logger.warn("Error parsing Span Sampling Rules \"#{e}\" for rules: #{rules}")
+                return nil
+              end
+
+              parse_list(list)
+            end
+
+            # Parses a list of Hashes containing the parsed JSON information
+            # for Single Span Sampling configuration.
+            # In case of parsing errors, `nil` is returned.
+            #
+            # @param rules [Array<String] the JSON configuration rules to be parsed
+            # @return [Array<Datadog::Tracing::Sampling::Span::Rule>] a list of parsed rules
+            # @return [nil] if parsing failed
+            def parse_list(rules)
+              unless rules.is_a?(Array)
+                # Using JSON terminology for the expected error type
+                Datadog.logger.warn("Span Sampling Rules are not an array: #{JSON.dump(rules)}")
+                return nil
+              end
+
+              parsed = rules.map do |hash|
+                unless hash.is_a?(Hash)
+                  # Using JSON terminology for the expected error type
+                  Datadog.logger.warn("Span Sampling Rule is not a key-value object: #{JSON.dump(hash)}")
```

Comment:
Minor, but in terms of debugability, would it be easier to perhaps just `.inspect` the wrong inputs?